cmd/jujud/agent/unit/manifolds.go

```diff
@@ -127,15 +153,57 @@ func Manifolds(config ManifoldsConfig) dependency.Manifolds {
 			LogSource:     config.LogSource,
 		}),
 
+		// The upgrade steps gate is used to coordinate workers which
+		// shouldn't do anything until the upgrade-steps worker has
+		// finished running any required upgrade steps. The flag of
+		// similar name is used to implement the isFullyUpgraded func
+		// that keeps upgrade concerns out of unrelated manifolds.
+		upgradeStepsGateName: gate.ManifoldEx(config.UpgradeStepsLock),
+		upgradeStepsFlagName: gate.FlagManifold(gate.FlagManifoldConfig{
+			GateName:  upgradeStepsGateName,
+			NewWorker: gate.NewFlagWorker,
+		}),
+
+		// The upgrade check gate is used to coordinate workers which
+		// shouldn't do anything until the upgrader worker has
+		// completed its first check for a new tools version to
+		// upgrade to. The flag of similar name is used to implement
+		// the isFullyUpgraded func that keeps upgrade concerns out of
```

Comment:
you should also copy across ifFullyUpgraded, and use it in migrationFortressName, to prevent migration occurring before the unit agent is upgraded

Reply:
good catch - done.